### PR TITLE
[AD] Implement AD annotations V2 for Pods, Service, Endpoints checks

### PR DESCRIPTION
**Pull request by juliogreff**
### What does this PR do?

This adds a new way of annotating K8s pods and services with AD annotations. Instead
of having to supply 3 different annotations (`check_names`, `init_configs`,
`instances`), users can now provide AD configs in a single `checks`
annotation as in the example below:

```
ad.datadoghq.com/foobar.checks: |
{
	"apache": {
		"instances": [
			{"apache_status_url":"http://%%host%%/server-status?auto2"}
		]
	}
}
```

This is also valid for service and endpoints checks (with `ad.datadoghq.com/service.checks` and `ad.datadoghq.com/endpoints.checks`, respectively).

### Describe how to test/QA your changes

Add V2 annotations to a pod and to a service (remember to add both endpoints and service annotations), and check that the checks are scheduled correctly. V1 annotations should remain working until there is a V2 annotation for the same container name, in which case the new config should take over.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: releasenotes/notes/ad-annotations-v2-a897fabaf00623c0.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+enhancements:
+  - |
+    The Agent now supports a single "ad.datadoghq.com/$IDENTIFIER.checks"
+    annotation in Kubernetes Pods and Services to configure Autodiscovery
+    checks. It merges the contents of the existing "check_names",
+    "init_configs" and "instances" annotations into a single JSON value.
```
Comment (kayayarai) [inline releasenotes/notes/ad-annotations-v2-a897fabaf00623c0.yaml:7]:
```suggestion
    "init_configs", and "instances" annotations into a single JSON value.
```

**Review by kayayarai — APPROVED** [review]

File: pkg/autodiscovery/common/utils/annotations.go [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+		} else {
+			configs = append(configs, c...)
+		}
+	}
+
+	return configs, errors
+}
+
+// parseChecksJSON parses an AD annotation v2
+// (ad.datadoghq.com/redis.check_names) JSON string into []integration.Config.
```
Comment (ahmed-mez) [inline pkg/autodiscovery/common/utils/annotations.go:145]:
```suggestion
// (ad.datadoghq.com/redis.checks) JSON string into []integration.Config.
```

File: pkg/autodiscovery/common/utils/annotations.go [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+func parseChecksJSON(adIdentifier string, checksJSON string) ([]integration.Config, error) {
+	var namedChecks map[string]struct {
+		Name       string           `json:"name"`
+		InitConfig integration.Data `json:"init_config"`
+		Instances  []interface{}    `json:"instances"`
+	}
+
+	err := json.Unmarshal([]byte(checksJSON), &namedChecks)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot parse check configuration: %w", err)
```
Comment (ahmed-mez) [inline pkg/autodiscovery/common/utils/annotations.go:155]:
```suggestion
		return nil, fmt.Errorf("cannot parse check configuration: %w", err)
```

File: pkg/autodiscovery/common/utils/annotations.go (hunk 2 above)
Comment (ahmed-mez) [reply pkg/autodiscovery/common/utils/annotations.go:145]: Let's also document that it fills `ADIdentifiers` and puts a default `InitConfig`

File: pkg/autodiscovery/common/utils/annotations.go [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+	return configs, errors
+}
+
+// parseChecksJSON parses an AD annotation v2
+// (ad.datadoghq.com/redis.check_names) JSON string into []integration.Config.
+func parseChecksJSON(adIdentifier string, checksJSON string) ([]integration.Config, error) {
+	var namedChecks map[string]struct {
+		Name       string           `json:"name"`
+		InitConfig integration.Data `json:"init_config"`
```
Comment (ahmed-mez) [inline pkg/autodiscovery/common/utils/annotations.go:149]: shouldn't this be of type `*integration.Data` to mark it as optional field? https://go.dev/play/p/1oxDEprdUG0

File: pkg/autodiscovery/common/utils/annotations.go [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+	}
+
+	return res, nil
+}
+
+// ParseJSONValue returns a slice of slice of ConfigData parsed from the JSON
+// contained in the `value` parameter
+func ParseJSONValue(value string) ([][]integration.Data, error) {
+	if value == "" {
+		return nil, fmt.Errorf("Value is empty")
```
Comment (ahmed-mez) [inline pkg/autodiscovery/common/utils/annotations.go:292]:
```suggestion
		return nil, fmt.Errorf("value is empty")
```

File: pkg/autodiscovery/common/utils/annotations_test.go [hunk 6]
```diff
@@ -0,0 +1,757 @@
...
+			annotations: map[string]string{
+				"ad.datadoghq.com/redis.checks": `{
+					"redisdb": {}
+				}`,
+				"service-discovery.datadoghq.com/redis.check_names": `["foo"]`,
+				"ad.datadoghq.com/redis.check_names":                `["bar"]`,
+			},
+			adIdentifier: "redis",
+			checkNames:   []string{"redisdb"},
+		},
```
Comment (ahmed-mez) [inline pkg/autodiscovery/common/utils/annotations_test.go:59]: let's add a test case with multiple check names?

File: releasenotes/notes/ad-annotations-v2-a897fabaf00623c0.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+---
+enhancements:
+  - |
+    The Agent now supports a single "ad.datadoghq.com/$IDENTIFIER.checks"
+    annotation in Kubernetes Pods and Services to configure Autodiscovery
+    checks. It merges the contents of the existing "check_names",
+    "init_configs", and "instances" annotations into a single JSON value.
```
Comment (ahmed-mez) [inline releasenotes/notes/ad-annotations-v2-a897fabaf00623c0.yaml:7]: should we also add a note that `init_configs` is optional in the new field?

**Review by ahmed-mez — APPROVED** [review]
few minor comments but LGTM!

**Pull request merged**